akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala

```diff
@@ -704,10 +704,29 @@ trait FlowOps[+Out, +Mat] {
    * '''Completes when''' upstream completes
    *
    * '''Cancels when''' downstream cancels
+   *
+   * See also [[FlowOps.scan]]
    */
   def fold[T](zero: T)(f: (T, Out) ⇒ T): Repr[T] = andThen(Fold(zero, f))
 
   /**
+   * Similar to `fold` but uses first element as zero element.
+   * Applies the given function towards its current and next value,
+   * yielding the next current value.
+   *
+   * '''Emits when''' upstream completes
+   *
+   * '''Backpressures when''' downstream backpressures
+   *
+   * '''Completes when''' upstream completes
+   *
+   * '''Cancels when''' downstream cancels
+   *
+   * See also [[FlowOps.fold]]
+   */
+  def reduce[T >: Out](f: (T, T) ⇒ T): Repr[T] = via(new Reduce[T](f))
```

Comment:
Unfortunately the stricter signature requires us to use `@uncheckedVariance` on `f` if we remove the T type parameter. I guess this is the reason for the signature of `Seq#reduce` as well—and I consider this a shortcoming of Scala’s declaration site variance approach (because it is impossible to prove to the compiler that this is not actually a problem here, although it could potentially be a problem elsewhere; the alternative of saying “shut up, compiler!” is not nice because it disables the check completely).

So, I’d say that we accept this potential broadening in the Scala API but restrict the Java API because there we can express what we actually want. That has the benefit of removing one potentially failing type inference for javac ;-)